File: 1-two-sum.java
class Solution {
    public int[] twoSum(int[] nums, int target) {       
        Map<Integer, Integer> valueIndexMap = new HashMap<>();
        for(int i = 0; i < nums.length; i++) {
            int num = nums[i];
            int supplement = target - num;
            if(valueIndexMap.containsKey(supplement)) {
                return new int[] { valueIndexMap.get(supplement), i};
            }
            
            valueIndexMap.put(num, i);         
        }
        
        return null;
    }
}

File: 1_Two_Sum.py
I scan the array once while storing each number's index in a hash map.
For each number, I compute its complement (target - value).
If the complement already exists in the map, I return the two indices immediately.
This gives O(n) time and O(n) space.

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        value_to_index = {}

        for index, value in enumerate(nums):
            complement = target - value
            if complement in value_to_index:
                return [value_to_index[complement], index]

            value_to_index[value] = index
        
        return [-1, -1]
